tests/core/data/test_network.py

```diff
@@ -0,0 +1,334 @@
+
+    assert network.shape[0] == network.adjoint().shape[1]
+    assert network.shape[1] == network.adjoint().shape[0]
+    assert network.dims[0] == network.adjoint().dims[1]
+    assert network.dims[1] == network.adjoint().dims[0]
+
+    np.testing.assert_almost_equal(node.tensor.conj(),
+                                   network.adjoint().nodes.pop().tensor)
+
+def test_transpose():
+    node = random_node((2,3))
+    network = Network([node[0]], [node[1]])
+
+    assert network.shape[0] == network.transopose().shape[1]
+    assert network.shape[1] == network.transopose().shape[0]
+    assert network.dims[0] == network.transopose().dims[1]
+    assert network.dims[1] == network.transopose().dims[0]
+
+    np.testing.assert_almost_equal(node.tensor,
+                                   network.transopose().nodes.pop().tensor)
+
+class TestContract():
+    node1 = random_node((2,))
+    node2 = random_node((2, 2, 2))
+    node1[0] ^ node2[2]
+
+    network = Network([node2[0]], [node2[1]], [node1, node2])
+
+    expect = (node1@node2).tensor
+    def test_default(self):
+        result = self.network.contract()
+        result = result.nodes.pop().tensor
+
+        np.testing.assert_allclose(result, self.expect)
+
+    def test_edge_order(self):
+        # We transpose the final result by changing the final_edge_order
+        result = self.network.contract(final_edge_order=self.network.in_edges +
+                                       self.network.out_edges)
+        result = result.nodes.pop().tensor
+
+        np.testing.assert_allclose(result, self.expect.T)
+
+def test_to_array():
+    node1 = random_node((2,))
+    node2 = random_node((2, 2, 2))
+    node1[0] ^ node2[2]
+
+    network = Network(node2[0:2], [], [node1, node2])
+    # We transpose the final result by changing the final_edge_order
+    result = network.to_array()
+    expect = (node1@node2).tensor.reshape((4,1))
+
+    np.testing.assert_allclose(result, expect)
+
+class TestMatmul:
+
+    @pytest.mark.parametrize("dim_in, dim_out",
+                            [([4], [2,2]),
+                             ([2,2], [4]),
+                             ([2,10], [2,5,2]),
+                             ([2,10], [2,2,5]),
+                             ([2], [2]),
+                            ])
+    def test_comatible(self, dim_in, dim_out):
+        """Tests that matrix multiplication works with networks that have
+        different dimensions. This test does not check for the correct tensor 
+        structure but rather that the output is numerically correct once
+        contracted."""
+        right = random_node(dim_out)
+        left = random_node(dim_in)
+
+        right_net = Network(right[:], [])
+        left_net = Network([], left[:])
+
+        result = left_net@right_net
+
+        # Desired
+        left = left.tensor.reshape((1, np.prod(dim_out)))
+        right = right.tensor.reshape((np.prod(dim_in), 1))
+        desired = left@right
```

Comment:
```suggestion
        desired = left @ right
```

Reply:
I will run black after committing all the suggestions.